File: backend/laravel/app/Services/GoogleTokenVerifier.php
<?php

namespace App\Services;

use Google_Client;
use Exception;

class GoogleTokenVerifier
{
    private $clientId;

    public function __construct()
    {
        $this->clientId = env('GOOGLE_CLIENT_ID');
    }

    public function verifyToken($idToken)
    {
        $client = new Google_Client(['client_id' => $this->clientId]);

        try {
            $payload = $client->verifyIdToken($idToken);
            if ($payload) {
                return $payload['email'];
            }
            throw new Exception('Invalid Google token.');
        } catch (Exception $e) {
            throw new Exception('Failed to verify Google token.');
        }
    }
}

File: backend/flask/app/services/google_token_verifier.py
from google.oauth2 import id_token
from google.auth.transport import requests
from config.settings import Config

class GoogleTokenVerifier:
    @staticmethod
    def verify_token(id_token_str):
        try:
            id_info = id_token.verify_oauth2_token(id_token_str, requests.Request(), Config.GOOGLE_CLIENT_ID)
            return id_info['email']
        except ValueError:
            raise Exception('Invalid Google token.')
